docs/DafnyRef/Attributes.md

```diff
@@ -472,21 +476,24 @@ This is mostly helpful for debugging which assertion is taking the most time to
 ### 22.2.22. `{:synthesize}` {#sec-synthesize-attr}
 
 The `{:synthesize}` attribute must be used on methods that have no body and
-return one or more fresh objects. During compilation, 
-the postconditions associated with such a
+return one or more fresh or mocked objects. To create a fresh instance of 
+an object, use the "fresh" parameter (`{:synthesize "fresh"}`). To create a 
+mocked instance of an object, use the "mock" parameter (`{:synthesize "mock"}`).
```

Comment:
My intention for suggesting the name `{:synthesize ... }` for this feature was to keep the semantics as abstract as possible: create an object whose behavior aligns with the specification. Using a mocking framework is just one possible implementation strategy, and one that could change in the future: alternatively you could actually synthesize source code in the target language that obeys the specification.

I'm concerned about committing to the semantics of `{:synthesize "fresh"}`, since AFAICT those semantics are to return a new instance of an object without invoking any constructor. Depending on the compilation strategy, that could easily create an object that causes errors/crashes/exceptions when used, rather than the intention (I believe) of producing a valid but underspecified object, whose fields are initialized arbitrarily.

It seems to me that it is only valid to use `{:synthesize "fresh"}` if `fresh(o)` is literally the only specified post-condition. Could you switch implementation strategies automatically based on that instead? Alternatively, could you stick with the mocking implementation instead, in the name of simplicity? (I recall we discussed this earlier but I can't recall the reason for wanting to make this distinction).

Reply:
I have updated this branch so that it matches the C# version of `synthesize`. I did this by using spies instead of mocks. By using spies, the need for a "fresh" argument is eliminated, as fields can be accessed and updated in a spy (they could not be with a mock). A spy allows for the same customization of behavior as a mock. The difference is that a spy comes with all the functionality of a real object by default, and that functionality can be changed, whereas a mock comes with none of the functionality of a real object by default, so everything must be added.

Reply:
Excellent, I'm so much happier with that! Thanks for sticking with this for the better outcome. :)

Please do revert this documentation change since it's not needed any more, aside from the Java specific remarks below.